dandischema/metadata.py

```diff
@@ -265,45 +265,116 @@ def validate(
 
 def migrate(
     obj: dict,
-    to_version: Optional[str] = DANDI_SCHEMA_VERSION,
+    to_version: str = DANDI_SCHEMA_VERSION,
     skip_validation: bool = False,
 ) -> dict:
-    """Migrate dandiset metadata object to new schema"""
-    obj = deepcopy(obj)
+    """
+    Migrate a Dandiset metadata instance to a newer DANDI schema version
+
+    :param obj: The Dandiset metadata instance to migrate
+    :param to_version: The target DANDI schema version to migrate to
+    :param skip_validation: If True, the given instance will not be validated before
+        migration. Otherwise, the instance will be validated before migration.
+    :return: The Dandiset metadata instance migrated to the target DANDI schema version
+    :raises ValueError: If the provided instance doesn't have a `schemaVersion` field
+        that specifies a valid DANDI schema version
+    :raises ValueError: If `to_version` is not a valid DANDI schema version or an
+        allowed target schema
+    :raises ValueError: If the target DANDI schema version is lower than the DANDI
+        schema version of the provided instance
+    """
+
+    # ATM, we only support the latest schema version as a target. See definition of
+    # `ALLOWED_TARGET_SCHEMAS` for details
     if len(ALLOWED_TARGET_SCHEMAS) > 1:
         raise NotImplementedError(
             "ATM code below supports migration to current version only"
         )
+
+    # --------------------------------------------------------------
+    # Validate DANDI schema version provided in the metadata instance
+    # --------------------------------------------------------------
+    # DANDI schema version of the provided instance
+    instance_ver = obj.get("schemaVersion")
+    if instance_ver is None:
+        msg = (
+            "The provided Dandiset metadata instance, `obj`, does not have a "
+            "`schemaVersion` field, for specifying DANDI schema version."
+        )
+        raise ValueError(msg)
+    if not isinstance(instance_ver, str):
+        msg = (
+            "The provided Dandiset metadata instance, `obj`, has a non-string "
+            "`schemaVersion` field, for specifying DANDI schema version."
+        )
+        raise ValueError(msg)
+    # Check if `instance_ver` is a valid DANDI schema version
+    try:
+        # DANDI schema version of the provided instance in tuple form
+        instance_ver_tuple = version2tuple(instance_ver)
+    except ValueError as e:
+        msg = (
+            "The provided Dandiset metadata instance, `obj`, has an invalid "
+            "`schemaVersion` field, for specifying DANDI schema version."
+        )
+        raise ValueError(msg) from e
+    if instance_ver not in ALLOWED_INPUT_SCHEMAS:
+        raise ValueError(f"Current input schemas supported: {ALLOWED_INPUT_SCHEMAS}.")
```

Comment:
```suggestion
        raise ValueError(f"Input schema version {instance_ver!r} not in supported input schemas: {ALLOWED_INPUT_SCHEMAS}")
```

Reply:
Suggestion taken and committed with modifications.